verilog/analysis/checkers/instance_shadow_rule_test.cc

```diff
@@ -0,0 +1,174 @@
+      {"module foo; logic a; class boo; endclass:boo; endmodule:foo;"},
+      {"module foo; logic a; begin boo; bit var1; end:boo; endmodule:foo;"},
+      {"module foo; logic a; begin boo;", "if(a) begin baz\n",
+       "$display(\"\t Value of i = %0d\",i);", "\nend:baz", "\nend:boo;",
+       "endmodule:foo;"},
+      {"module foo; logic a; begin boo;",
+       "loop: for(int i=0;i<5;i++) begin baz\n",
+       "$display(\"\t Value of i = %0d\",i);", "\nend:baz", "\nend:boo;",
+       "endmodule:foo;"},
+      {
+          "module foo;\n",
+          "genvar i;\n",
+          "generate\n",
+          "initial begin\n",
+          "for(int i=0;i<5;i++) $display(\"\t i = %d\",i);\n",
+          "end\n",
+          "endgenerate\n",
+          "endmodule:foo;\n",
+      },
+      {
+          "module foo;\n",
+          "initial begin\n",
+          "int a;\n",
+          "end\n",
+          "initial begin\n",
+          "int a;\n",
+          "end\n",
+          "endmodule:foo;\n",
+      },
+      {
+          "module foo;\n",
+          "initial begin\n",
+          "int a;\n",
+          "end\n",
+          "int a;\n",
+          "endmodule:foo;\n",
+      },
+  };
+  RunLintTestCases<VerilogAnalyzer, InstanceShadowRule>(
+      kInstanceShadowingTestCases);
+}
+
+TEST(InstanceShadowingTest, FunctionFailures) {
+  constexpr int kToken = SymbolIdentifier;
+  const std::initializer_list<LintTestCase> kInstanceShadowingTestCases = {
+      {"module foo; logic ", {kToken, "foo"}, "; endmodule:foo;"},
+      {"module foo; logic a; logic ", {kToken, "a"}, "; endmodule:foo;"},
+      {"module foo; logic a; class ",
+       {kToken, "foo"},
+       "; endclass:foo; endmodule:foo;"},
+      {"module foo; logic a; class boo; logic ",
+       {kToken, "boo"},
+       "; endclass:boo; endmodule:foo;"},
+      {"module foo; logic a; class boo; logic ",
+       {kToken, "foo"},
+       "; endclass:boo; endmodule:foo;"},
+      {
+          "module foo;\n",
+          "int i;\n",
+          "initial begin\n",
+          "for(int ",
+          {kToken, "i"},
+          "=0;i<5;i++) $display(\"i = %d\",i);\n",
+          "end\n",
+          "endmodule:foo;\n",
+      },
+      {
+          "module foo;\n",
+          "int a;\n",
+          "initial begin\n",
+          "int ",
+          {kToken, "a"},
+          ";\nend\n",
+          "endmodule:foo;\n",
+      },
+      {
+          "module foo;\n",
+          "initial begin\n",
+          "int a;\n",
+          "int ",
+          {kToken, "a"},
+          ";\n",
+          "end\n",
+          "endmodule:foo;\n",
+      },
+      {"function automatic int foo (input bit in);\n",
+       "bit ",
+       {kToken, "in"},
+       ";\n",
+       "endfunction\n"},
+      {"function automatic int foo (input bit in, input bit ",
+       {kToken, "in"},
+       ");\n",
+       "bit out;\n",
+       "endfunction\n"},
+  };
+  RunLintTestCases<VerilogAnalyzer, InstanceShadowRule>(
```

Comment:
What should happen with test cases that resemble K&R style declarations, like:

```systemverilog
module foo(a, b);
input wire a;   // not a shadow, but attaches a type to the `a` port
output reg b;  // not a shadow
endmodule
```

Can you include a few cases like this?

Reply:
It actually does complain about this situation. I added a test that confirms that, and a TODO to fix.
Given that by default, the rule is disabled and that this case of ancient code is rare (errm, should be), I'd like to defer that to a later PR.